### PR TITLE
Add file/line information to main error type with macros

### PR DESCRIPTION
**Pull request by svix-gabriel**
This closes https://github.com/svix/svix-webhooks/issues/574, by adding diagnostic information to the main error type along with some helpful macros.

## Motivation

For internal server errors, we want to visibility into both the underlying error, in addition to where the error originated from. e.g. for code like this
```rust
fn foo() -> crate::error::Result<()> {
   let app = Application.fetch(...)?;
   app.bar = 12;
   app.save()?;
   Ok(())
}
```
Just looking at the logs, it can be hard to diagnose whether the underlying db error came from `Application.fetch(...)?` or `app.save()?`.

We want to be able to quickly correlate an error to a specific line, while still keeping error handling relatively boilerplate free.

## Solution

First, we extend the `error::Error` type with some diagnostic information - literally just the line number and file name.

To avoid really boilerplate heavy code, most of the `error::Error` initialization logic is now handled by macros. `queue_err!`, `generic_err!`, etc. This makes it pretty easy to add the file/line information without making error initialization excessively verbose (in some cases, we're actually able to make error initialization easier than before).

One point of friction is that we can't easily attach `location` information when propagating errors using `?` - which we lean on pretty heavily when converting `DbErr`, `RedisError`, `bb8::RunError`, etc. Adding the location information to an error should be easy, but it should also be something that is difficult to forget to do.

To circumvent this problem, I dropped the `From<...> for Error` implementations for `DbErr`, `RedisError`, etc. These were replaced with a very similar trait to just transforms `std::result::Result<T, E>` into `crate::error::Result<T>` for `DbErr`, `RedisError`, etc. There's then a `ctx!` macro that handles the conversion, and adds the `location` information. Now error handling looks closer too

```rust
fn foo() -> crate::error::Result<()> {
   let app = ctx!(Application.fetch(...))?;
   app.bar = 12;
   ctx!(app.save())?;
   Ok(())
}
```

So, marginally more boilerplate. But forgetting to include `ctx!` now becomes a compiler error, forcing developers to add the diagnostic information using `ctx!`.

In addition, `ctx!` adds tracing information if the error is already a `crate::result::Result`. 

What does this end up looking like in practice? If you have code like this
```rust
fn fallible_a() -> std::result::Result<(), DbErr> {
    Err(DbErr::Custom("db borked!".into()))
}

fn fallible_b() -> Result<()> {
    ctx!(fallible_a()) // line 182 in application.s
}

// . . .

ctx!(fallible_b())?; // line 190 in application.rs
```
when the error is transformed into a response from the axum handlers, we get logs like
```
2022-09-14T14:13:04.156518Z ERROR HTTP request{...}: svix_server::error: type: Database("Custom Error: db borked!"), location: ["svix-server/src/v1/endpoints/application.rs:182", "svix-server/src/v1/endpoints/application.rs:190"]
```

### CODE REVIEW

File: server/svix-server/src/core/cryptography.rs [hunk 1]
```diff
@@ -8,7 +8,8 @@ use chacha20poly1305::{Key, XChaCha20Poly1305, XNonce};
 use ed25519_compact::*;
 use rand::Rng;
 
-use crate::error::{Error, Result};
+use crate::error::Result;
+use crate::generic_err;
```
Comment (svix-gabriel) [inline server/svix-server/src/core/cryptography.rs:12]: I couldn't figure out how to scope the macros in `error.rs` in the actual `error` module, so we could call `error::generic!(...)` instead of `generic_err!(...)`. Suggestions there would be helpful.

File: server/svix-server/src/error.rs [hunk 2]
```diff
@@ -54,53 +72,130 @@ impl error::Error for Error {
...
 
-impl<T: error::Error + 'static> From<bb8::RunError<T>> for Error {
-    fn from(err: bb8::RunError<T>) -> Self {
-        Error::Queue(err.to_string())
-    }
+/// Returns a [&'static str] of the file.rs\[<line_number>\].
+#[macro_export]
+macro_rules! location {
+    () => {
+        concat!(file!(), "[", line!(), "]")
```
Comment (tasn) [inline server/svix-server/src/error.rs:94]:
```suggestion
        concat!(file!(), ":", line!())
```
More standard?

File: server/svix-server/src/error.rs [hunk 3]
```diff
@@ -16,35 +16,53 @@ use serde_json::json;
...
-    Queue(String),
-    /// Database error
-    Validation(String),
-    /// Any kind of HttpError
-    Http(HttpError),
+#[derive(Debug)]
+pub struct Error {
+    // the file name and line number of the error. Used for debugging non Http errors
+    pub location: &'static str,
+    pub error_type: ErrorType,
```
Comment (tasn) [inline server/svix-server/src/error.rs:23]: Nitpick:
```suggestion
    pub type_: ErrorType,
```
or `t` or `typ`. The word error is redundant.

File: server/svix-server/src/error.rs [hunk 4]
```diff
@@ -16,35 +16,53 @@ use serde_json::json;
...
-    /// Queue error
-    Queue(String),
-    /// Database error
-    Validation(String),
-    /// Any kind of HttpError
-    Http(HttpError),
+#[derive(Debug)]
+pub struct Error {
+    // the file name and line number of the error. Used for debugging non Http errors
+    pub location: &'static str,
```
Comment (tasn) [inline server/svix-server/src/error.rs:22]: Correct me if I'm wrong, but we are not maintaining the error chain? E.g. we are missing location (and type) of the caller? Maybe it makes sense to have a linked-list type of thing of vectors (or heck, a pre-allocated vec of size 10 that should always be sufficient) so we keep the trace?

Comment (tasn) [reply server/svix-server/src/error.rs:22]: Every time we call `ctx!` btw, obviously it'll be lost if we don't.

Comment (svix-gabriel) [reply server/svix-server/src/error.rs]: Yeah, good catch.

Should be updated now so that `ctx!` also adds the tracing (I ended up renaming the `IntoSvixError` trait, hopefully it makes a bit more sense now).

PR description should also be updated showing what the end "trace" result looks like.

I didn't go through and add `ctx!(...)` to _every_ single relevant callsite, though I can if you want (it'll be a bit tedious). Unfortunately, I'm not sure how to make forgetting to call `ctx!` on a `crate::error::Result` a compiler error like I was when forgetting to call `ctx!` on a normal `Result`.

File: server/svix-server/src/core/cryptography.rs (hunk 1 above)
Comment (tasn) [reply server/svix-server/src/core/cryptography.rs:12]: Oh right, macros are always crate-wide?

So maybe at least prefix it with err? So `err_generic` so it's at least namespaced like this, even if poorly.

File: server/svix-server/src/error.rs (hunk 4 above)
Comment (tasn) [reply server/svix-server/src/error.rs]: Nah, no need to add it everywhere for now.

Btw, I remember you preferred `e!`, why the change? (I don't care much either way)

Comment (svix-gabriel) [reply server/svix-server/src/error.rs]: Haha, I still _personally_ prefer `e!` just for brevity. But I also think `ctx!` is more descriptive / it's easier to infer what `ctx!` is doing.

Maybe not though 🤷.

Comment (tasn) [reply server/svix-server/src/error.rs]: I think I agree. :)

File: server/svix-server/src/core/cryptography.rs [hunk 5]
```diff
@@ -74,7 +74,7 @@ impl Encryption {
             let nonce = XNonce::from_slice(&nonce);
             let mut ciphertext = cipher
                 .encrypt(nonce, data)
-                .map_err(|_| crate::error::Error::Generic("Encryption failed".to_string()))?;
+                .map_err(|_| crate::err_generic!("Encryption failed"))?;
```
Comment (svix-daniel) [inline server/svix-server/src/core/cryptography.rs:77]: Is the `crate::` prefix necessary here (or line 93)?

File: server/svix-server/src/core/types.rs [hunk 6]
```diff
@@ -471,7 +471,7 @@ impl EndpointSecretMarker {
         let encrypted = (v & Self::ENCRYPTED_FLAG) != 0;
         let v = v & !Self::ENCRYPTED_FLAG;
         let type_ = EndpointSecretType::try_from(v)
-            .map_err(|_| crate::error::Error::Generic("Invalid marker value".to_string()))?;
+            .map_err(|_| crate::err_generic!("Invalid marker value"))?;
```
Comment (svix-daniel) [inline server/svix-server/src/core/types.rs:474]: Is there a reason not to `use crate::err_generic` and get rid of the prefix? I imagine that macro name is pretty unique. (Also lines 545 and 609)

File: server/svix-server/src/error.rs [hunk 7]
```diff
@@ -16,35 +16,59 @@ use serde_json::json;
...
-    Generic(String),
-    /// Database error
-    Database(String),
-    /// Queue error
-    Queue(String),
-    /// Database error
-    Validation(String),
-    /// Any kind of HttpError
-    Http(HttpError),
+#[derive(Debug)]
```
Comment (svix-daniel) [inline server/svix-server/src/error.rs:19]: Not a big concern, but how does the debug print look like in the terminal? It might make more sense to implement `std::fmt::Debug` instead of deriving it?

Should there be alternative formats depending on log format?

Again, not super important, at least for the first draft.

File: server/svix-server/src/error.rs [hunk 8]
```diff
@@ -54,53 +78,139 @@ impl error::Error for Error {
...
-    fn from(err: DbErr) -> Error {
-        if let DbErr::Query(ref err_str) = err {
-            // Surely there's a better way
-            if err_str.contains("duplicate key value violates unique constraint") {
-                Error::Http(HttpError::conflict(None, None))
+impl<T> Traceable<T> for std::result::Result<T, DbErr> {
+    fn trace(self, location: &'static str) -> Result<T> {
+        self.map_err(|err| {
+            let typ = if let DbErr::Query(ref err_str) = err {
+                // Surely there's a better way
```
Comment (svix-daniel) [inline server/svix-server/src/error.rs:162]: Without changes to SeaORM, I don't think there is one. I'd probably remove the comment.

**Comment by svix-daniel** [comment]
Overall, these changes are pretty good. It's a shame that the syntax can't be nicer, but I understand it'd require a lot more effort to get it to look more natural.

One question:
> In addition, ctx! adds tracing information if the error is already a crate::result::Result.
Is there any way to enforce adding a `ctx!` invocation easily? It'd be useful to get the traceback as bulky as possible (within reason), so I'd want to get it to as close to the top as possible in its use.

I have one really hacky idea for that, but it'd be terribly hard to do, so I'll leave the topic be if there's no obvious solution.

File: server/svix-server/src/core/types.rs (hunk 6 above)
Comment (tasn) [reply server/svix-server/src/core/types.rs:474]: I think having some prefix is nice. Because we'll also have `database!` which is less unique...

Comment (svix-gabriel) [reply server/svix-server/src/core/types.rs]: I didn't have a specific reason to use the `crate::`, other than it's what was already there. Removed.

File: server/svix-server/src/core/cryptography.rs (hunk 5 above)
Comment (svix-gabriel) [reply server/svix-server/src/core/cryptography.rs]: Nope (same reason as before). Removed

**Comment by svix-gabriel** [comment]
> Is there any way to enforce adding a ctx! invocation easily?

Yeah, I definitely agree having some kind of compile time enforcement for calling `ctx!` would be nice. I can't think of any reasonable approach though

File: server/svix-server/src/core/otel_spans.rs [hunk 9]
```diff
@@ -128,7 +128,7 @@ impl<B> MakeSpan<B> for AxumOtelSpanCreator {
         );
 
         if let Some(key) = idempotency_key {
-            span.record("idempotency_key", &key);
+            span.record("idempotency_key", key);
```
Comment (tasn) [inline server/svix-server/src/core/otel_spans.rs:131]: Why all of these unrelated changes?

Comment (svix-gabriel) [reply server/svix-server/src/core/otel_spans.rs:131]: I was getting clippy failures in the build https://github.com/svix/svix-webhooks/actions/runs/3136424848/jobs/5093319099#step:8:727

I can remove these diffs, but I didn't think the PR would be mergable unless they passed

Comment (svix-gabriel) [reply server/svix-server/src/core/otel_spans.rs]: Fixed

**Review by svix-daniel — APPROVED** [review]
